### ISSUE TITLE
Early exaggeration

### ISSUE THREAD

**Issue by clane9** [karpathy/tsnejs#15]
Super late but I think there is a small bug with how early exaggeration is implemented. [The code](https://github.com/karpathy/tsnejs/blob/13ece5d7e751a5180a59ca396013e800e1059c20/tsne.js#L351) has:

```javascript
var premult = 4 * (pmul * P[i*N+j] - Q[i*N+j]) * Qu[i*N+j]; 
```

whereas I think it should be

```javascript
var premult = 4 * pmul * (P[i*N+j] - Q[i*N+j]) * Qu[i*N+j];
```

I.e. the early exaggeration factor `pmul` should multiply the overall gradient rather than `P` alone.

The t-SNE authors describe early exaggeration as scaling `P` for the first few iterations. Since `P` is a constant, you'd think you could achieve the effect just by scaling `P` in the gradient. This is what the code does.

But at the same time, since the loss (`-(P * Q.log()).sum()`) is multiplied by `P`, scaling `P` should also scale the overall gradient. Now I'm confused.

In Appendix A of the [original paper](https://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf), the authors assume the sum of `P` is 1. But under early exaggeration the sum is `pmul`. In this case, the `q_ij` term in the gradient becomes `pmul * q_ij`. So really the overall gradient should be scaled by `pmul`.

Not sure if this "fix" should be implemented, since the code has worked reliably for years. But thought it was worth noting.

**Comment by clane9** [comment]
I realized this is a deliberate feature, not a bug. Every t-SNE implementation includes it. And disabling it often dramatically hurts optimization performance (even if you try to compensate with a larger learning rate). In fact, there is some [theoretical work](https://arxiv.org/pdf/1803.01768.pdf) establishing some convergence guarantees for t-SNE with early exaggeration. Still, it was interesting to me that this is not exactly gradient descent.
